### PR TITLE
feat: add lab5

### PR DESCRIPTION
**Pull request by s1rserg**
## *Added Database Setup*
- Installed PostgreSQL as RDBMS
- Created tables using BIGSERIAL for auto-incrementing primary keys
- Added test data for voting system demonstration
- Implemented via migrations using Liquibase

## *Added Core Database Features*
- Implemented sequences for primary key generation using BIGSERIAL
- Created tables:
  - `voting` (main voting information)
  - `candidate` (voting options)
  - `vote` (user votes tracking)
- Added CASCADE deletion for referential integrity

## *Rewrote Repository Layer*
- Created DAO interfaces and implementations for:
  - VotingRepository
  - CandidateRepository
  - VoteRepository
- Implemented CRUD operations
- Added custom search methods:
  - Find all votings (was already implemented)
  - Find votes by voting id and user id
  - Find candidates by voting

## *Rewrote Service Layer*
- Implemented business logic interfaces and classes:
  - VotingService
  - CandidateService
  - VoteService
- Added transactional voting process
- Implemented vote counting and validation

## *Rewrote Controller Layer*
- Added RESTful endpoints:
  - PATCH `/api/votings/{id}/votes` - update a vote
  - DELETE `/api/votings/{id}/votes` - retract a vote

## Entity Relationship Diagram
```mermaid
erDiagram
    voting ||--o{ candidate : "has"
    voting ||--o{ vote : "receives"
    candidate ||--o{ vote : "receives"

    voting {
        bigint id PK
        varchar(255) title
        text description
        boolean active
        bigint creator_user_id
    }

    candidate {
        bigint id PK
        varchar(255) name
        integer votes
        bigint voting_id FK
    }

    vote {
        bigint id PK
        bigint voting_id FK
        bigint candidate_id FK
        bigint user_id
    }
```

### CODE REVIEW

File: src/main/java/com/example/votingsystem/repository/impl/VotingDAOImpl.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    public List<Voting> getAll(String title, int page, int size) {
+        StringBuilder sql = new StringBuilder("SELECT * FROM voting");
+        if (title != null && !title.isEmpty()) {
+            sql.append(" WHERE title LIKE ?");
+        }
+        sql.append(" LIMIT ? OFFSET ?");
+        int offset = (page - 1) * size;
```
Comment (HoroshkoMykhailo) [inline src/main/java/com/example/votingsystem/repository/impl/VotingDAOImpl.java:46]: page по дефолту 0, воно викликає помилку через те, що offset тоді від'ємний

File: src/main/java/com/example/votingsystem/controller/VotingController.java [hunk 2]
```diff
@@ -29,7 +32,7 @@ public ResponseEntity<?> getAllVotings(
             @RequestParam(defaultValue = "0") int page,
```
Comment (HoroshkoMykhailo) [inline src/main/java/com/example/votingsystem/controller/VotingController.java]: Тут поставити дефолтне значення 1

**Review by HoroshkoMykhailo — CHANGES_REQUESTED** [review]

**Review by HoroshkoMykhailo — APPROVED** [review]

File: src/main/java/com/example/votingsystem/service/impl/VoteServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    @Override
+    public List<Vote> getByVotingId(Long votingId) {
+        return voteDAO.getByVotingId(votingId);
+    }
+
+    @Override
+    @Transactional
+    public boolean retract(Long votingId, Long userId) {
+        Vote vote = voteDAO.getByVotingAndUserId(votingId, userId).orElseThrow(() -> new IllegalArgumentException("No vote found for the given user in this voting"));
+        voteDAO.deleteByVotingAndUserId(vote.getId(), userId);
```
Comment (HoroshkoMykhailo) [inline src/main/java/com/example/votingsystem/service/impl/VoteServiceImpl.java]: Тут має бути votingId, а не voteId
`voteDAO.deleteByVotingAndUserId(votingId, userId);`

**Review by HoroshkoMykhailo — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/votingsystem/repository/impl/VotingDAOImpl.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public Voting update(Voting voting) {
+        String sql = "UPDATE voting SET title = ?, description = ?, active = ?, creator_user_id = ? WHERE id = ?";
+        jdbcTemplate.update(sql, voting.getTitle(), voting.getDescription(), voting.isActive(), voting.getCreatorUserId(), voting.getId());
+        return getById(voting.getId()).orElseThrow(() -> new RuntimeException("Failed to retrieve updated voting record"));
+    }
+
+    @Override
+    public void delete(Long id) {
+        String sql = "DELETE FROM voting WHERE id = ?";
```
Comment (HoroshkoMykhailo) [inline src/main/java/com/example/votingsystem/repository/impl/VotingDAOImpl.java:64]: Ну і тут змінити
```suggestion
        String sql = "DELETE FROM voting WHERE user_id = ?";
```

**Review by HoroshkoMykhailo — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/votingsystem/repository/impl/VotingDAOImpl.java (hunk 4 above)
Comment (s1rserg) [reply src/main/java/com/example/votingsystem/repository/impl/VotingDAOImpl.java:64]: Тут трохи не зрозумів

Comment (HoroshkoMykhailo) [reply src/main/java/com/example/votingsystem/repository/impl/VotingDAOImpl.java:64]: не то, ти все правильно зробив

**Review by HoroshkoMykhailo — APPROVED** [review]

File: src/main/java/com/example/votingsystem/repository/impl/VotingDAOImpl.java (hunk 4 above)
Comment (HoroshkoMykhailo) [reply src/main/java/com/example/votingsystem/repository/impl/VotingDAOImpl.java:64]: Я файли перепутав

**Pull request merged**